[dev-launcher][dev-menu][ios] Add support for Fabric

gabrieldonadel opened the pull request:
# Why

Closes ENG-7955

This PR  adds dev-launcher support for the new architecture. 
 
# How

 - Updates `expo-dev-launcher` and `expo-dev-menu` podspecs to support `NEW_ARCH_ENABLED` flags
 - Add `EXDevLauncherBridgeDelegate` class that inherits from `RCTAppDelegate` and is responsible for creating the root view and bridge 
 - Replace `DevMenuRCTCxxBridgeDelegate`with `DevMenuRCTAppDelegate`
- Fix `EXDevLauncherBridgeDelegate` not able to relaunch DevLauncher 
- Add `ExpoDevLauncherBridgeDelegateHandler` to handle opening apps from `ExpoDevLauncherReactDelegateHandler`
- Update `openDevMenuFromReactNative` to ensure the DevMenu is closed before opening the react-native dev menu


# Test Plan

Run `fabric-tester` and `bare-expo` on iOS

 
<table>
    <tr><th>fabric-tester</th><th>bare-expo</th></tr>
    <tr>
    <td>
        <video src="https://github.com/expo/expo/assets/11707729/1176f862-4459-4a39-8c11-16f73b2dfe46"/>
   </td>
   <td>
   <video src="https://github.com/expo/expo/assets/11707729/5f603af5-d970-4144-ade6-0943b829d225"   />  
    </td>
</tr> 
</table> 

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

gabrieldonadel committed:
[dev-launcher][ios] Add support for Fabric

gabrieldonadel committed:
Partially fix app launch

gabrieldonadel committed:
Simplify EXDevLauncherBridgeDelegate

gabrieldonadel committed:
Simplify podspec

gabrieldonadel committed:
Fix lauching DevMenu

gabrieldonadel commented on packages/expo-dev-menu/ios/DevMenuRCTBridge.mm:
```diff
@@ -106,25 +119,38 @@ - (void)reloadWithReason:(NSString *)reason {}
+    self.turboModuleManager = [[RCTTurboModuleManager alloc] initWithBridge:bridge
+                                                                  delegate:(id<RCTTurboModuleManagerDelegate>)self
+                                                                  jsInvoker:bridge.jsCallInvoker];
+    std::unique_ptr<facebook::react::JSExecutorFactory> executorFactory = RCTAppSetupDefaultJsExecutorFactory(bridge, self.turboModuleManager);
+
+    #if __has_include(<reacthermes/HermesExecutorFactory.h>)
+        auto rawExecutorFactory = executorFactory.get();
+        auto hermesExecFactory = dynamic_cast<facebook::react::HermesExecutorFactory*>(rawExecutorFactory);
+        if (hermesExecFactory != nullptr) {
+            hermesExecFactory->setEnableDebugger(false);
+        }
+    #endif
+
+    return executorFactory;
+  #endif
```
@Kudo do you think it's fine to disable the debugger like this?

Kudo replied:
this is intentional. hermes debugger does not well support loading different bundles. the latest loading bundle will override the previous one from debugging. we had an issue in dev-client when dev-menu shown, the debugger target will be overridden as dev-menu's bundle. people should reload the app to override the debugging target back to the app.
with the `setEnableDebugger` we can prevent the problem.

gabrieldonadel replied:
But do you think this casting from `JSExecutorFactory` to `HermesExecutorFactory` would still work as intended?
I'm trying to keep the `RCTAppSetupDefaultJsExecutorFactory(bridge, self.turboModuleManager)` call

Kudo replied:
thanks for the effort to use the `RCTAppSetupDefaultJsExecutorFactory`. most react-native code requires `-frtti` so it should be okay to do the dynamic_cast here. moreover, you could also try static_cast if that works, given the fact that inside `#if __has_include(<reacthermes/HermesExecutorFactory.h>)` block, the executor should be hermes.

gabrieldonadel committed:
Prevent DevLauncherController from removing fabric root view

gabrieldonadel committed:
Fix imports

gabrieldonadel committed:
Enable DevMenu to run on Fabric

gabrieldonadel commented on packages/expo-dev-launcher/ios/EXDevLauncherController.m:
```diff
@@ -564,12 +568,18 @@ - (void)onAppContentDidAppear
   [[NSNotificationCenter defaultCenter] removeObserver:self name:RCTContentDidAppearNotification object:nil];
 
   dispatch_async(dispatch_get_main_queue(), ^{
+    #ifdef RCT_NEW_ARCH_ENABLED
+      #define EXPECTED_ROOT_VIEW RCTSurfaceView
+    #else
+      #define EXPECTED_ROOT_VIEW RCTRootView
+    #endif
```
RootView class name changes when running Fabric

gabrieldonadel committed:
Ensure we use DevMenuRCTCxxBridge

gabrieldonadel commented on packages/expo-dev-launcher/ios/EXDevLauncherBridgeDelegate.mm:
```diff
@@ -0,0 +1,74 @@
+}
+
+- (RCTRootView *)createRootViewWithModuleName:(NSString *)moduleName launchOptions:(NSDictionary * _Nullable)launchOptions application:(UIApplication *)application{
+    BOOL enableTM = NO;
+#if RCT_NEW_ARCH_ENABLED
+    enableTM = YES;
+#endif
+
+    RCTAppSetupPrepareApp(application, enableTM);
+
+    if (!self.bridge) {
+        self.bridge = [self createBridgeWithDelegate:self launchOptions:launchOptions];
+    }
+
+#ifdef RCT_NEW_ARCH_ENABLED
+    _contextContainer = std::make_shared<facebook::react::ContextContainer const>();
+    _reactNativeConfig = std::make_shared<facebook::react::EmptyReactNativeConfig const>();
+    _contextContainer->insert("ReactNativeConfig", _reactNativeConfig);
+    self.bridgeAdapter = [[RCTSurfacePresenterBridgeAdapter alloc] initWithBridge:self.bridge
+                                                                 contextContainer:_contextContainer];
+    self.bridge.surfacePresenter = self.bridgeAdapter.surfacePresenter;
+#endif
+
+    NSMutableDictionary *initProps = [NSMutableDictionary new];
+#ifdef RCT_NEW_ARCH_ENABLED
+    initProps[kRNConcurrentRoot] = @YES;
+#endif
+
+
+    return [super createRootViewWithBridge:self.bridge moduleName:moduleName initProps:initProps];
```
Based on https://github.com/facebook/react-native/blob/main/packages/react-native/Libraries/AppDelegate/RCTAppDelegate.mm

gabrieldonadel replied:
Yeah, unfortunately I don't think we can call `(BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions` to reuse that code. Let me create a linear task to propose updating RCTAppDelegate

gabrieldonadel replied:
Here it is https://linear.app/expo/issue/ENG-8940/propose-a-change-into-rctappdelegate-to-make-it-more-flexible-for

gabrieldonadel replied:
Sure, I've just updated it to use the new 0.72 code and added a `DevClientAppDelegate` class that is used between dev-launcher and dev-menu so we can keep all the copied code in one single place

gabrieldonadel committed:
Fix jsExecutorFactoryForBridge

gabrieldonadel committed:
Update RootView class name

gabrieldonadel committed:
Add changelog entry

Kudo commented on apps/fabric-tester/ios/fabrictester.xcodeproj/project.pbxproj:
```diff
@@ -355,8 +355,18 @@
+					"@executable_path/Frameworks",
+				);
 				MARKETING_VERSION = 1.0;
+				OTHER_CPLUSPLUSFLAGS = (
+					"$(inherited)",
+					"-DRCT_NEW_ARCH_ENABLED=1",
+					"-DFOLLY_NO_CONFIG",
+					"-DFOLLY_MOBILE=1",
+					"-DFOLLY_USE_LIBCPP=1",
+				);
```
are the changes intentional? does user need to change these in their project?

gabrieldonadel replied:
Ops, these changes are not necessary, I've just reverted it

gabrieldonadel committed:
Update packages/expo-dev-launcher/expo-dev-launcher.podspec

Co-authored-by: Kudo Chien <kudo@expo.dev>

gabrieldonadel committed:
Disable swiftlint implicitly_unwrapped_optional warning on DevMenuAppInstance

gabrieldonadel committed:
[dev-launcher][dev-menu][ios] Add support for Fabric - Part 2 (#22821)

Closes ENG-7955

This PR is Part 2 of a series of PRs that will add dev-launcher support
for the new architecture.

Part 1 #22184

- Fix `EXDevLauncherBridgeDelegate` not able to relaunch DevLauncher
- Add `ExpoDevLauncherBridgeDelegateHandler` to handle opening apps from
`ExpoDevLauncherReactDelegateHandler`
- Update `openDevMenuFromReactNative` to ensure the DevMenu is closed
before opening the react-native dev menu

Run `fabric-tester` and `bare-expo` on iOS

<table>
    <tr><th>fabric-tester</th><th>bare-expo</th></tr>
    <tr>
    <td>
<video
src="https://github.com/expo/expo/assets/11707729/69e80f31-3d8e-4135-afd7-0f4264052a79"/>
   </td>
   <td>
<video
src="https://github.com/expo/expo/assets/11707729/99b10d00-45b3-4b3e-b7c0-2cc7599ebad7"
/>
    </td>
</tr>
</table>

<!--
Please check the appropriate items below if they apply to your diff.
This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg:
https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).

Kudo commented on packages/expo-dev-menu/ios/DevMenuRCTBridge.h:
```diff
@@ -16,7 +23,9 @@ NS_ASSUME_NONNULL_BEGIN
-@interface DevMenuRCTCxxBridgeDelegate : NSObject
+@interface DevMenuRCTAppDelegate : RCTAppDelegate
+
+- (void)createBridgeWithAdapter:(NSDictionary *_Nullable)launchOptions;
```
nit: naming thing: is that intentional to pass `launchOptions` as `adapter`?

gabrieldonadel replied:
Ohh, not really, I've just renamed it to `createBridgeAndSetAdapterWithLaunchOptions`

gabrieldonadel commented on packages/expo-dev-menu/ios/DevMenuDevOptionsDelegate.swift:
```diff
@@ -24,7 +24,9 @@ class DevMenuDevOptionsDelegate {
     DevMenuManager.shared.hideMenu()
 
-    bridge?.requestReload()
+    DispatchQueue.main.async {
+      RCTTriggerReloadCommandListeners("Dev menu - reload")
+    }
```
`requestReload` is deprecated 
https://github.com/facebook/react-native/blob/main/packages/react-native/React/Base/RCTBridge.m#L325-L331

gabrieldonadel committed:
Fix SwiftLint

gabrieldonadel commented on packages/expo-dev-menu/ios/DevMenuRCTBridge.mm:
```diff
@@ -27,10 +27,36 @@
 #import <reacthermes/HermesExecutorFactory.h>
 #endif
 
+#import "RCTAppDelegate+Private.h"
+#import "RCTAppSetupUtils.h"
```
I'm actually not able to import using `#import <React/RCTAppSetupUtils.h>`

<img width="283" alt="image" src="https://github.com/expo/expo/assets/11707729/958bf910-a7ff-48ee-8742-f8ae42b7fb0e">

gabrieldonadel replied:
Fixed

gabrieldonadel commented on packages/expo-dev-menu/ios/DevMenuRCTBridge.mm:
```diff
@@ -27,10 +27,36 @@
+#import <ReactCommon/RCTTurboModuleManager.h>
+#import <react/config/ReactNativeConfig.h>
+#import <react/renderer/runtimescheduler/RuntimeSchedulerCallInvoker.h>
+#import "RCTLegacyInteropComponents.h"
```
Same thing as `<React/RCTAppSetupUtils.h>`, for some reason build fails with `'React/RCTLegacyInteropComponents.h' file not found`

gabrieldonadel committed:
Address PR comments

gabrieldonadel commented on packages/expo-dev-menu/ios/RCTAppDelegate+Private.h:
```diff
@@ -0,0 +1,11 @@
+@interface RCTAppDelegate (DevMenuRCTAppDelegate)
+
+#ifdef __cplusplus
+- (std::unique_ptr<facebook::react::JSExecutorFactory>)jsExecutorFactoryForBridge:(RCTBridge *)bridge;
```
Sure, I've just moved it to DevMenuRCTBridge.mm

gabrieldonadel commented on packages/expo-dev-menu/ios/DevMenuAppInstance.swift:
```diff
@@ -38,9 +38,9 @@ class DevMenuAppInstance: DevMenuRCTCxxBridgeDelegate, RCTBridgeDelegate {
+  // MARK: RCTAppDelegate
 
-  func sourceURL(for bridge: RCTBridge!) -> URL! {
+  @objc(sourceURLForBridge:) override func sourceURL(for bridge: RCTBridge!) -> URL! {
```
That was something that Xcode was suggesting but does not seem necessary, removed.

gabrieldonadel commented on packages/expo-dev-launcher/ios/EXDevLauncherController.m:
```diff
@@ -266,12 +274,8 @@ - (void)navigateToLauncher
-                                                   moduleName:@"main"
-                                            initialProperties:@{}];
+  UIApplication *application = [UIApplication sharedApplication];
+  UIView *rootView = [_bridgeDelegate createRootViewWithModuleName:@"main" launchOptions:_launchOptions application:application];
```
Updated

gabrieldonadel committed:
Rename createBridgeWithAdapter to createBridgeAndSetAdapterWithLaunchOptions

gabrieldonadel committed:
Add DevClientAppDelegate and update code to latest 0.72

gabrieldonadel commented on packages/expo-dev-menu/ios/DevMenuRCTBridge.mm:
```diff
@@ -106,25 +132,46 @@ - (void)reloadWithReason:(NSString *)reason {}
+    _contextContainer->insert("ReactNativeConfig", _reactNativeConfig);
+    self.bridgeAdapter = [[RCTSurfacePresenterBridgeAdapter alloc] initWithBridge:self.bridge
+                                                                 contextContainer:_contextContainer];
+    self.bridge.surfacePresenter = self.bridgeAdapter.surfacePresenter;
```
updated

gabrieldonadel committed:
Revert unnecessary xcode project changes

Kudo commented on packages/expo-dev-launcher/ios/EXDevLauncherBridgeDelegate.mm:
```diff
@@ -0,0 +1,74 @@
+#import "EXDevLauncherController.h"
+
+#import <React/RCTBundleURLProvider.h>
+#import "RCTAppSetupUtils.h"
```
sorry it should be `#import <React-RCTAppDelegate/RCTAppSetupUtils.h>`

gabrieldonadel replied:
Thanks, that fixes it

Kudo commented on packages/expo-dev-launcher/ios/expo-dev-launcher-Bridging-Header.h:
```diff
@@ -13,3 +13,4 @@
 #import "EXDevLauncherRedBoxProtocol.h"
 #import "EXDevLauncherRedBox.h"
 #import "RCTBundleURLProvider+Private.h"
+#import "RCTAppSetupUtils.h"
```
```suggestion
```

remove unused import

gabrieldonadel replied:
removed

Kudo commented on packages/expo-dev-menu/ios/Headers/EXDevMenu-Bridging-Header.h:
```diff
@@ -17,6 +17,7 @@
 #import <React/RCTLogBox.h>
 #import <React/RCTRedBox.h>
 #import <React/RCTPackagerConnection.h>
+#import <React-RCTAppDelegate/RCTAppDelegate.h>
```
```suggestion
```

remove unused import

gabrieldonadel replied:
removed

Kudo reviewed (commented):
great! the code is now much more clear and promising. just some small cleanup and should be good to go 🚀

gabrieldonadel committed:
Add EXRCTAppDelegateInterceptor

gabrieldonadel committed:
Update RCTAppDelegate imports

gabrieldonadel commented on packages/expo-dev-menu/ios/DevMenuRCTBridge.mm:
```diff
@@ -27,10 +27,15 @@
 #import <reacthermes/HermesExecutorFactory.h>
 #endif
 
+#import "RCTAppSetupUtils.h"
```
Updated

gabrieldonadel commented on packages/expo-dev-menu/ios/DevMenuAppInstance.swift:
```diff
@@ -53,15 +52,20 @@ class DevMenuAppInstance: DevMenuRCTCxxBridgeDelegate, RCTBridgeDelegate {
+
+  func concurrentRootEnabled() -> Bool {
+    return true
+  }
```
RCTAppDelegate used to use that but it's no longer necessary, removed

gabrieldonadel commented on packages/expo-dev-launcher/expo-dev-launcher.podspec:
```diff
@@ -43,11 +41,22 @@ Pod::Spec.new do |s|
     'DEFINES_MODULE' => 'YES',
     'OTHER_CFLAGS[config=Debug]' => other_c_flags,
     'OTHER_SWIFT_FLAGS[config=Debug]' => other_swift_flags,
+    "HEADER_SEARCH_PATHS" => "\"$(PODS_TARGET_SRCROOT)/ReactCommon\" \"$(PODS_TARGET_SRCROOT)\" \"$(PODS_ROOT)/RCT-Folly\" \"$(PODS_ROOT)/boost\" \"$(PODS_ROOT)/DoubleConversion\" \"$(PODS_ROOT)/Headers/Private/React-Core\"",
```
Updated

gabrieldonadel commented on packages/expo-dev-menu/ios/DevClientAppDelegate.mm:
```diff
@@ -0,0 +1,84 @@
+
+#import <React/RCTBundleURLProvider.h>
+#import <React/RCTRuntimeExecutorFromBridge.h>
+#import "RCTAppSetupUtils.h"
```
Updated

Kudo commented on packages/expo-dev-launcher/ios/EXRCTAppDelegateInterceptor.h:
```diff
@@ -0,0 +1,15 @@
+
+@end
+
+NS_ASSUME_NONNULL_END
```
what's the difference of this and the `packages/expo-modules-core/ios/ReactDelegates/EXRCTBridgeDelegateInterceptor.h`? could we just keep one?

gabrieldonadel replied:
The only difference is that this inherits from `DevClientAppDelegate`, that way we can call `createBridgeAndSetAdapterWithLaunchOptions` inside `ExpoDevLauncherReactDelegateHandler`

gabrieldonadel committed:
Move RCTAppDelegate interface to DevClientAppDelegate.mm

gabrieldonadel commented on packages/expo-dev-menu/ios/DevClientAppDelegate.h:
```diff
@@ -0,0 +1,21 @@
+#import <React_RCTAppDelegate/RCTAppDelegate.h>
+#endif
+
+@interface RCTAppDelegate (DevClientAppDelegate)
+
+- (void)unstable_registerLegacyComponents;
+
+@end
```
Sure, moved

gabrieldonadel commented on packages/expo-dev-menu/app/hooks/useDevSettings.tsx:
```diff
@@ -88,8 +88,7 @@ export function useDevSettings() {
 
   const openRNDevMenu = React.useCallback(async () => {
-    await DevMenu.openDevMenuFromReactNative();
-    DevMenu.closeMenu();
```
Not necessary anymore, we have to close this from the native side to ensure we first remove the expo-dev-menu overlay and then show the DevMenu from React Native otherwise both menus get closed. 

E.g.

https://github.com/expo/expo/assets/11707729/f7cea942-0b28-4716-89db-d4bfc3ceed43


That's the reason for the changes on packages/expo-dev-menu/ios/Modules/DevMenuInternalModule.swift

Kudo commented on packages/expo-dev-menu/ios/DevClientAppDelegate.mm:
```diff
@@ -0,0 +1,90 @@
+
+#import <React/RCTBundleURLProvider.h>
+#import <React/RCTRuntimeExecutorFromBridge.h>
+#import <React-RCTAppDelegate/RCTAppSetupUtils.h>
```
```suggestion
#if __has_include(<React_RCTAppDelegate/RCTAppSetupUtils.h>)
// for importing the header from framework, the dash will be transformed to underscore
#import <React_RCTAppDelegate/RCTAppSetupUtils.h>
#else
#import <React-RCTAppDelegate/RCTAppSetupUtils.h>
#endif
```

gabrieldonadel replied:
updated

Kudo commented on packages/expo-dev-menu/ios/DevClientAppDelegate.mm:
```diff
@@ -0,0 +1,90 @@
+
+#import <react/renderer/runtimescheduler/RuntimeScheduler.h>
+#import <react/renderer/runtimescheduler/RuntimeSchedulerCallInvoker.h>
+#import <React-RCTAppDelegate/RCTAppDelegate.h>
```
```suggestion
#if __has_include(<React_RCTAppDelegate/RCTAppDelegate.h>)
// for importing the header from framework, the dash will be transformed to underscore
#import <React_RCTAppDelegate/RCTAppDelegate.h>
#else
#import <React-RCTAppDelegate/RCTAppDelegate.h>
#endif
```

Kudo commented on packages/expo-dev-menu/ios/DevMenuRCTBridge.mm:
```diff
@@ -27,10 +27,15 @@
 #import <React/RCTCxxBridgeDelegate.h>
 #import <React/RCTJSIExecutorRuntimeInstaller.h>
+#import <React/RCTRuntimeExecutorFromBridge.h>
+#import <React-RCTAppDelegate/RCTAppSetupUtils.h>
```
```suggestion
#if __has_include(<React_RCTAppDelegate/RCTAppSetupUtils.h>)
// for importing the header from framework, the dash will be transformed to underscore
#import <React_RCTAppDelegate/RCTAppSetupUtils.h>
#else
#import <React-RCTAppDelegate/RCTAppSetupUtils.h>
#endif
```

Kudo reviewed (commented):
sorry found some problem from my previous suggestion.
after that it's good to go.

gabrieldonadel committed:
Update RCTAppDelegate imports

Kudo reviewed (approved):
🔥🚀🚀

gabrieldonadel committed:
Fix imports when fabric is disabled

gabrieldonadel committed:
Fix JSCExecutorFactory header import
